### ISSUE TITLE
Discrepancy/Bug in the fNTorecFN operation

### ISSUE THREAD

**Issue by Axionix77** [bsg-external/HardFloat#28]
We suspect a possible bug in the recoding module for double-precision inputs. The bug appears to be in the fNTorecFN module. The two inputs a and b that are being compared in the attached JasperGold snapshot are not equal, and our testbench also computes that they are not equal. However, the reconversion (recoded) module appears to make them identical after the recoding. We were expecting to see a.out and b.out to mirror a and b in terms of staying unequal.

<img width="1025" alt="Screenshot 2024-11-18 at 16 29 25" src="https://github.com/user-attachments/assets/4a0798f8-ea1c-475b-b54e-c300b570f418">

**Comment by dpetrisko** [comment]
Interesting, that does seem odd. Those are two distinct subnormal numbers so they should map to distinct recodings

This is not our code though, we’re mirroring:
https://www.jhauser.us/arithmetic/HardFloat.html

> Bugs in HardFloat and other comments can be reported to me at jh@jhauser.us.

**Comment by jhauser-us** [comment]
> This is not our code though

The bug was introduced by modifications made here.  It's not a bug in the original.  So yes, it is your code.

**Comment by taylor-bsg** [comment]
@dpetrisko

**Comment by dpetrisko** [comment]
Well the only modification to that file is the CLZ

https://github.com/bsg-external/HardFloat/issues/10

so presumably that discrepancy is the source

**Comment by jhauser-us** [comment]
I wrote:
> The bug was introduced by modifications made here.

Actually, I should be more careful and say I don't know that for certain.  What I do know is that the error is occurring for them because module `bsg_counting_leading_zeros` within `fNToRecFN` is returning the wrong value.  As should be clear from the name, that's not a module from the original HardFloat sources.  Hence, I'm kicking the ball back to you guys.

**Comment by dpetrisko** [comment]
@jhauser-us yes same conclusion here. Sorry for implying it was yours, I hastily thought that was an unmodified file!

**Comment by Axionix77** [comment]
Once the patch has been applied, we can re-test this in formal to confirm the fix.

**Comment by dpetrisko** [comment]
So does this bug still exist with #30 applied? And confirmed it doesn’t exist with upstream master ?(bf579553c927092650f5976d8da573af1ac64741)

**Comment by Axionix77** [comment]
Hi, 
This bug does exist in the most recent version, however it does not appear in the bf57955 version.

**Comment by dpetrisko** [comment]
Great thanks I will check it out

**Comment by dpetrisko** [comment]
[testcase.tar.gz](https://github.com/user-attachments/files/18648053/testcase.tar.gz)

Hello @Axionix77 I cannot reproduce this bug either.

I've attached a test case with the following output:

```
[                   0] Initializing testbench...
[                   8] nReset raised
[                 100] test input: 0000000200000003 (0000000000000000000000000000001000000000000000000000000000000011)
[                 110] test output: 03ef0000000180000 (00011111011110000000000000000000000000000000110000000000000000000)
	expected: 03ef0000000180000 (00011111011110000000000000000000000000000000110000000000000000000)
		DUT CLZ in: 0000200000003 count: 12
		Hardfloat CLZ in: 0000200000003 count: 12
		BSG CLZ in: 0000200000003 count: 12
[                 120] test input: 0000000000000003 (0000000000000000000000000000000000000000000000000000000000000011)
[                 130] test output: 03cf8000000000000 (00011110011111000000000000000000000000000000000000000000000000000)
	expected: 03cf8000000000000 (00011110011111000000000000000000000000000000000000000000000000000)
		DUT CLZ in: 0000000000003 count: 32
		Hardfloat CLZ in: 0000000000003 count: 32
		BSG CLZ in: 0000000000003 count: 32
[                 140] All tests finished; exiting

- testbench.sv:100: Verilog $finish
```

input 1 0000000200000003 -> 03ef0000000180000
input 2 0000000000000003 -> 03cf8000000000000

Can you double check your setup and close this and #31 if they are also not legitimate?

**Comment by dpetrisko** [comment]
Hi @Axionix77 any update on this?

**Comment by Axionix77** [comment]
Hi, 
Sorry for the delay.
So I checked everything again. Here is the update:
The current files are all fine, I only get the error when I run things with an older version of the bsg files from another repo. 
Thank you.
